### ISSUE TITLE
resizing volatile volumes

### ISSUE THREAD

**Issue by xaki23** [QubesOS/qubes-issues#5645]
**The problem you're addressing (if any)**
a user on freenode #qubes brought up the topic of "resizing volatile volume", and the results were ... unexpected. 

**Describe the solution you'd like**
i would prefer to "qvm-volume resize blahvm:volatile 123GiB" to work just as it does on :private or :root

**Where is the value to a user, and who might that user be?**
consistent behavior of qvm tooling.
flexible space management.
the requesting user on irc mainly wanted to use it for actual swap, my usecase tends to be temp storage space.

**Describe alternatives you've considered**
manually creating tempfiles in dom0, looping them, and attaching works. (see fedora version upgrade guide)
using space inside private or root works too.
all these options have manual overhead, need cleanup, or have unexpected longer-term impact.

**Additional context**
will describe some individual findings below in comments.
they are more "things we tried" and "partial workarounds" stuff. 

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**
not aware of really on-topic documentation, we ended up reading quite a bit of code though.

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**
not aware of any.

**Comment by xaki23** [comment]
```
$ qvm-create -l gray -P lvm test-vol
$ qvm-volume info test-vol:volatile | grep size
size               10737418240
$ qvm-volume resize test-vol:volatile 123GiB
  Logical volume vm-test-vol-volatile not found in volume group qubes_dom0.
$ qvm-run -p test-vol "grep 'xvdc$' /proc/partitions"
 202       32   10485760 xvdc
$ qvm-volume resize test-vol:volatile 123GiB
Online resize of volume qubes_dom0/vm-test-vol-volatile failed (you need to resize filesystem manually): Automatic resize of 'volatile' not supported
$ qvm-run -p test-vol "grep 'xvdc$' /proc/partitions"
 202       32  128974848 xvdc
$
```

so resizing while the vm is halted doesnt work at all (more on that later).
resizing while the vm is running sortof-works, but has the usual "fighting hotplug while reformating volumes" issues.
--> sortof works, but not in a persisted way, and with some nuisance details

**Comment by xaki23** [comment]
so after figuring out the "not found" error in the last comment is actualy from "lvextend", there is an obvious way to sidestep it:
```
# qvm-create -l gray -P lvm test-vol
# lvcreate -V1G -T qubes_dom0/pool00 -n vm-test-vol-volatile
  Logical volume "vm-test-vol-volatile" created.
# qvm-volume resize test-vol:volatile 123GiB
# qvm-run -p test-vol "grep 'xvdc' /proc/partitions"
 202       32  128974848 xvdc
 202       33    1048576 xvdc1
 202       35  127925248 xvdc3
# 
```
so if i create the volume first, then resize, then start ... it works mostly like expected.
the volume has the right size, so the initrd picks up on it too and does a proper partitioning.
--> kinda works, but in an "ugly workaround" way.

the "volume not found" error on halted-resize is caught here: https://github.com/QubesOS/qubes-core-admin/blob/0a66a0c7dd801d475b6ea346e3677a72f705a917/qubes/storage/lvm.py#L798
so one workaround for this might be to add an "if volume==volatile and error matches not-found: return True" here. 

but, there is another aspect to this, and i didnt see this one coming ... continuing from the situation above:
```
# qvm-volume info test-vol:volatile | grep size
size               132070244352
# qvm-shutdown --wait test-vol
# qvm-volume info test-vol:volatile | grep size
size               10737418240
```
so the volume config reverts to default on shutdown.
so its still not persisted. 
and i cant find the code that does this "revert my volume config".

**Comment by xaki23** [comment]
so where does this 10GB default come from anyways?
afaict it is defaulted in per-vm-type like here: https://github.com/QubesOS/qubes-core-admin/blob/39a9e4e42243487e6039f78ffb4fc95e03b7c66a/qubes/vm/templatevm.py#L84
... meaning it uses a global default root volume size ...
which is defined here: https://github.com/QubesOS/qubes-core-admin/blob/39a9e4e42243487e6039f78ffb4fc95e03b7c66a/qubes/config.py#L76

so when changing that default in config.py and restarting qubesd or rebooting ... that actualy seems to change the default size of volatiles (and root) going forward.
i dont think i have to explain why that is not really anywhere near a solution.

**Comment by xaki23** [comment]
so, where to go from here?
i would prefer the "qvm-volume resize blah:volatile 1234" to work just as on any other volume.
how?
afaict the only parts missing is to make whatever is resetting the volume on vm shutdown respect the config and adding some mild ignorance for the resize op on halted.

@marmarek where is that reset-on-shutdown hiding?
and got any better ideas?

**Comment by marmarek** [comment]
Each vm volume has `size` attribute. It just happens that LVM driver use it rarely and prefers to use actual LVM volume size. Specifically, resize only resizes actual LVM volume, but don't update `size` attribute.
This is fixed in linked commit.
The `volume not found` error is also fixed, in another commit.
I've opened a PR with those two changes. I'd be glad for some testing.

Remaining issue is failed resize2fs call ("Online resize of volume qubes_dom0/vm-test-vol-volatile failed"), and this is slightly more complex. There is `qubes.ResizeDisk` qrexec service in a VM that receives volume name as an argument. It has handling only for `root` and `private` volumes right now (it needs special per-volume handling because of different partitioning etc).
I don't think that's a big deal for this use case, though, as you repartition that extra space manually anyway.

**Comment by xaki23** [comment]
status summary with 4.0.50:
- resize-enlarge while vm is offline works, vm will get larger volatile on next start and partition correctly
- resize-enlarge while vm is running changes the configured size, but doesnt change the size of the existing volatile volume. it is not just missing the repartition+resize inside the vm, even the plain xvdc blockdevice doesnt change size, so there is nothing to be done/repartitioned/resized inside the vm. the volume will have the correct size after restarting the vm
- resize-shrink doesnt work with the usual "not supported, use lvresize" message, which doesnt work in this case since there is nothing to lvresize (while stopped) or not persisted (if running).

in summary: works for me. thanks!
